File: TPII-2025-2/PadraoGoF/Estrutural/Flyweight/Flyweight01.py
# Flyweight Factory
class FabricaFlyweight:
    def __init__(self):
        self.flyweights = {}
    
    def obterFlyweight(self, id):
        if id not in self.flyweights:
            self.flyweights[id] = ProdutoFlyweight(id)
        return self.flyweights[id]

# Flyweight
class ProdutoFlyweight:
    def __init__(self, id):
        self.id = id
    
    def exibirDetalhes(self, nome, preco):
        print(f"ID: {self.id} | Produto:{nome:15} | R${preco:.2f}")

# Cliente:
class Cliente:
    def __init__(self):
        self.fabricaFlyweight = FabricaFlyweight()
        self.carrinho = []
    
    def addProduto(self, id, nome, preco):
        flyweight = self.fabricaFlyweight.obterFlyweight(id)
        self.carrinho.append({"flyweight": flyweight, "nome": nome, "preco": preco})
    
    def exibirCarrinho(self):
        print("Itens do Carrinho:")
        for item in self.carrinho:
            item["flyweight"].exibirDetalhes(item["nome"], item["preco"])

# Uso
cliente = Cliente()
cliente.addProduto("001", "Camiseta     ", 39.90)
cliente.addProduto("002", "Calça Jeans  ", 89.90)
cliente.addProduto("001", "Camiseta     ", 59.90)
cliente.exibirCarrinho()

File: TPII-2025-2/PadraoGoF/Estrutural/Flyweight/Flyweight01.js
// Flyweight Factory
class FabricaFlyeight{
    constructor(){
        this.flyweights = {};
    }

    obterFlyeight(id){
        if(!this.flyweights[id]){
            this.flyweights[id] = new ProdutoFlyeight(id);
        }
        return this.flyweights[id];
    }
}

// Flyeight
class ProdutoFlyeight{
    constructor(id){
        this.id = id;
    }

    exibirDetalhe(nome, preco){
        console.log(`ID: ${this.id} | R$${preco.toFixed(2)} | PROD: ${nome}`);
    }
}

// Cliente - USO DO PADRAO:
class Cliente{
    constructor(){
        this.fabricaFlyeight = new FabricaFlyeight();
        this.pedido = []
    }

    addProduto(id, nome, preco){
        const flyweights = this.fabricaFlyeight.obterFlyeight(id);
        this.pedido.push({flyweights, nome, preco});
    }

    exibirPedido(){
        console.log("Itens do Pedido:");
        this.pedido.forEach(item => {
            item.flyweights.exibirDetalhe(item.nome, item.preco);
        })
    }
}

// Aplicação Geral da Estrutura:
let valor = 99.99;
const cliente = new Cliente();

cliente.addProduto("001", "Camiseta   ", 39.90);
cliente.addProduto("002", "Calca Jeans", 89.90);
cliente.addProduto("003", "Bermuda    ", 39.90);
//---
cliente.addProduto("002", "Chinelo    ", valor);
cliente.addProduto("003", "Bone       ", 39.90);

cliente.exibirPedido();